packages/salesforcedx-vscode-core/test/jest/commands/deployExecutor.test.ts

```diff
@@ -0,0 +1,185 @@
+  let workspaceContextGetInstanceSpy: jest.SpyInstance;
+  let getUsernameStub: jest.SpyInstance;
+  let createSourceTrackingSpy: jest.SpyInstance;
+  let deploySpy: jest.SpyInstance;
+  let loadCacheStub: jest.SpyInstance;
+  let createDiffsStub: jest.SpyInstance;
+  let handleConflictsStub: jest.SpyInstance;
+
+  class TestDeployExecutor extends DeployExecutor<{}> {
+    constructor(s: string, t: string, x?: boolean) {
+      super(s, t);
+    }
+
+    protected getComponents(
+      response: ContinueResponse<{}>
+    ): Promise<ComponentSet> {
+      return new Promise(resolve => resolve(new ComponentSet()));
+    }
+  }
+
+  beforeEach(async () => {
+    jest.spyOn(process, 'cwd').mockReturnValue(dummyProcessCwd);
+    jest.spyOn(fs, 'existsSync').mockReturnValue(true);
+    workspaceContextGetInstanceSpy = jest
+      .spyOn(WorkspaceContext, 'getInstance')
+      .mockReturnValue(mockWorkspaceContext);
+    getUsernameStub = jest
+      .spyOn(ConfigUtil, 'getUsername')
+      .mockResolvedValue(dummyUsername);
+    createSourceTrackingSpy = jest
+      .spyOn(SourceTrackingService, 'createSourceTracking')
+      .mockResolvedValue({
+        ensureLocalTracking: ensureLocalTrackingSpy
+      } as any);
+    deploySpy = jest
+      .spyOn(dummyComponentSet, 'deploy')
+      .mockResolvedValue({ pollStatus: jest.fn() } as any);
+    loadCacheStub = jest
+      .spyOn(MetadataCacheService.prototype, 'loadCache')
+      .mockResolvedValue(dummyCacheResult);
+    createDiffsStub = jest
+      .spyOn(TimestampConflictDetector.prototype, 'createDiffs')
+      .mockReturnValue(dummyDiffs);
+    handleConflictsStub = jest
+      .spyOn(TimestampConflictChecker.prototype, 'handleConflicts')
+      .mockResolvedValue({} as any);
+  });
+
+  it('should create Source Tracking and call ensureLocalTracking before deploying', async () => {
+    // Arrange
+    deploySpy = jest
+      .spyOn(dummyComponentSet, 'deploy')
+      .mockResolvedValue({ pollStatus: jest.fn() } as any);
+    const executor = new TestDeployExecutor(
+      'testDeploy',
+      'force_source_deploy_with_sourcepath_beta'
+    );
+    (executor as any).setupCancellation = jest.fn();
+
+    // Act
+    await (executor as any).doOperation(dummyComponentSet, {});
+
+    // Assert
+    expect(createSourceTrackingSpy).toHaveBeenCalled();
+    expect(ensureLocalTrackingSpy).toHaveBeenCalled();
+    expect(deploySpy).toHaveBeenCalled();
+    const createSourceTrackingCallOrder =
+      createSourceTrackingSpy.mock.invocationCallOrder[0];
+    const ensureLocalTrackingSpyCallOrder =
+      ensureLocalTrackingSpy.mock.invocationCallOrder[0];
+    const deployCallOrder = deploySpy.mock.invocationCallOrder[0];
+    // In order to be sure that a Source Tracking instance is initialized
+    // and tracking files appropriately, create and ensureLocalTracking
+    // need to be called before the deploy operation is started.
+    expect(createSourceTrackingCallOrder).toBeLessThan(deployCallOrder);
```

Comment:
So this is just a way of looking at the order in which these are called, and making sure that createSourceTracking and ensureLocalTracking come before deployCallOrder? 

Also, is the 3rd assert a dupe of the first?

Reply:
Yep exactly.  And yes, third one was unneeded.  Good catch! Removed.